### PR TITLE
DM-21688: Create multi plugins for DiaObject summary statistics

### PR DESCRIPTION
**Pull request by morriscb**

### CODE REVIEW

File: python/lsst/ap/association/diaCalculationPlugins.py [hunk 1]
```diff
@@ -633,16 +657,21 @@ def calculate(self,
...
-                (fluxes[1:] - fluxes[:-1]) / (times[1:] - times[:-1]))
-        else:
-            self.fail(diaObject,
-                      ["{}{}".format(filterName, colName)
-                       for colName in self.outputCols])
+
+        def _maxSlope(df):
+            tmpDf = df[~np.logical_or(np.isnan(df["psFlux"]),
+                                      np.isnan(df["midPointTai"]))]
+            if len(tmpDf) <= 1:
```
Comment (ebellm) [inline python/lsst/ap/association/diaCalculationPlugins.py]: should be `< 2`

File: python/lsst/ap/association/diaCalculationPlugins.py [hunk 2]
```diff
@@ -633,16 +657,21 @@ def calculate(self,
...
+                return np.nan
+            times = tmpDf["midPointTai"].to_numpy()
+            timeArgs = times.argsort()
+            times = times[timeArgs]
+            fluxes = tmpDf["psFlux"].to_numpy()[timeArgs]
+            return ((fluxes[1:] - fluxes[:-1]) /
```
Comment (ebellm) [inline python/lsst/ap/association/diaCalculationPlugins.py]: suggest using `np.diff` here

File: tests/test_dia_calculation_plugins.py [hunk 3]
```diff
@@ -190,33 +294,42 @@ def testCalculate(self):
...
+        diaObjects = pd.DataFrame({"diaObjectId": [objId]})
+        fluxes = np.linspace(-1, 1, n_sources)
+        fluxes[4] = np.nan
+        diaSources = pd.DataFrame(
+            data={"diaObjectId": n_sources * [objId],
+                  "filterName": n_sources * ["r"],
+                  "diaSourceId": np.arange(n_sources, dtype=int),
+                  "psFlux": fluxes,
+                  "psFluxErr": np.ones(n_sources)})
+        run_multi_plugin(diaObjects, diaSources, "r", plug)
```
Comment (ebellm) [inline tests/test_dia_calculation_plugins.py]: test a non-zero value

File: tests/test_dia_calculation_plugins.py [hunk 4]
```diff
@@ -379,29 +525,36 @@ def testCalculate(self):
...
-        plug.calculate(diaObject, diaSources, diaSources, "r")
+        diaObjects = pd.DataFrame({"diaObjectId": [objId]})
+        diaSources = pd.DataFrame(
+            data={"diaObjectId": n_sources * [objId],
+                  "filterName": n_sources * ["r"],
+                  "diaSourceId": np.arange(n_sources, dtype=int),
+                  "psFlux": fluxes,
+                  "psFluxErr": np.ones(n_sources)})
+        run_multi_plugin(diaObjects, diaSources, "r", plug)
         cutFluxes = fluxes[~np.isnan(fluxes)]
```
Comment (ebellm) [inline tests/test_dia_calculation_plugins.py]: use the nan cut in `skew` itself

File: tests/test_dia_calculation_plugins.py [hunk 5]
```diff
@@ -410,31 +563,38 @@ def testCalculate(self):
...
 
-        diaObject = dict()
+        # Test expected MinMax fluxes with a nan set.
         fluxes[4] = np.nan
-        diaSources = pd.DataFrame(data={"psFlux": fluxes,
-                                        "psFluxErr": np.ones(n_sources)})
-        plug.calculate(diaObject, diaSources, diaSources, "r")
-        self.assertEqual(diaObject["rPSFluxMin"], -1)
-        self.assertEqual(diaObject["rPSFluxMax"], 1)
+        fluxes = np.linspace(-1, 1, n_sources)
```
Comment (ebellm) [inline tests/test_dia_calculation_plugins.py]: delete this line so it doesn't overwrite the `nan`

File: tests/test_dia_calculation_plugins.py [hunk 6]
```diff
@@ -443,39 +603,51 @@ def testCalculate(self):
         """Test flux maximum slope.
         """
         n_sources = 10
-        diaObject = dict()
+        objId = 0
+
+        # Test max slope value.
         fluxes = np.linspace(-1, 1, n_sources)
         times = np.linspace(0, 1, n_sources)
```
Comment (ebellm) [inline tests/test_dia_calculation_plugins.py]: append a value so the max slope is different than the slope of all the other points?

File: python/lsst/ap/association/diaCalculationPlugins.py [hunk 7]
```diff
@@ -430,15 +452,14 @@ def calculate(self,
...
-            diaObject["{}PSFluxChi2".format(filterName)] = np.nansum(
-                (delta / filterDiaSources["psFluxErr"]) ** 2)
-        else:
-            self.fail(diaObject,
-                      ["{}{}".format(filterName, colName)
-                       for colName in self.outputCols])
+        meanName = "{}PSFluxMean".format(filterName)
+
+        def _chi2(df):
+            delta = (df["psFlux"] - diaObjects.at[df.index[0][0], meanName])
```
[the remark this answers is not in the document]
Comment (morriscb) [reply python/lsst/ap/association/diaCalculationPlugins.py]: `df.diaObjectId.iat[0]` seems to be the way to go.

File: tests/test_dia_calculation_plugins.py (hunk 3 above)
Comment (morriscb) [reply tests/test_dia_calculation_plugins.py]: Realized there was a bug in the mean calculation where totWeight would be incorrect if a flux was nan but not the err. Added a test for both weighted means.

File: python/lsst/ap/association/association.py [hunk 8]
```diff
@@ -497,20 +513,63 @@ def match(self, dia_objects, dia_sources, score_struct):
...
+            ``?PSFluxData``
+                Number of data points used to calculate point source flux
+                summary statistics in each bandpass (`int`).
+        """
+        new_dia_object = {"diaObjectId": objId,
+                          "pmParallaxNdata": 0,
+                          "nearbyObj1": 0,
+                          "nearbyObj2": 0,
+                          "nearbyObj3": 0}
+        for f in ["u", "g", "r", "i", "z", "y"]:
```
[the remark this answers is not in the document]
Comment (morriscb) [reply python/lsst/ap/association/association.py]: We'll deal with this in DM-21267 and/or related tickets.

**Pull request merged**